include/tvm/runtime/container.h

```diff
@@ -1410,10 +1344,114 @@ inline String& String::operator=(std::string other) {
 
 inline String& String::operator=(const char* other) { return operator=(std::string(other)); }
 
-inline String operator+(const std::string lhs, const String& rhs) {
-  return lhs + rhs.operator std::string();
+template <typename T, typename U,
+          typename = typename std::enable_if<std::is_same<T, String>::value ||
+                                             std::is_same<T, std::string>::value>::type,
+          typename = typename std::enable_if<std::is_same<U, String>::value ||
+                                             (std::is_same<U, std::string>::value &&
+                                              !std::is_same<T, U>::value)>::type>
+inline String operator+(const T& lhs, const U& rhs) {
```

Comment:
The match is a bit too broad, let us still directly enumerate the combinations that involves String